HPCC-31511 Avoid keeping the workunit graphs in memory

richardkchapman opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Compiled

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman committed:
HPCC-31511 Avoid keeping the workunit graphs in memory

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday commented on esp/platform/CMakeLists.txt:
```diff
@@ -42,6 +42,7 @@ include_directories (
          ./../../system/jlib 
          ./../bindings/http/platform 
          ./../../common/environment 
+         ./../../common/workunit
```
I would have expected a need to add workunit to the link libraries as well.

richardkchapman replied:
Not sure why it builds without - probably because one of the libraries included already links to it. But I will add.

ghalliday commented on common/workunit/workunit.hpp:
```diff
@@ -1788,5 +1788,6 @@ extern WORKUNIT_API void executeThorGraph(const char * graphName, IConstWorkUnit
 extern WORKUNIT_API TraceFlags loadTraceFlags(IConstWorkUnit * wu, const std::initializer_list<TraceOption> & y, TraceFlags dft);
 
 extern WORKUNIT_API bool executeGraphOnLingeringThor(IConstWorkUnit &workunit, unsigned wfid, const char *graphName);
+extern WORKUNIT_API bool workunitGraphCacheEnabled;
```
I'm not a fan of exported variables.  We tend to provide get()/set() functions everywhere else except for roxie.  I may not care enough to avoid merging though.

richardkchapman replied:
If you can come up with a good reason, I don't mind changing. But there are plenty of other exported variables around...

ghalliday replied:
There are very few, except for roxie.  (E.g. I don't think there are any in this file.)
It generates slightly worse code (an extra indirect) - not significant in this case, but in general worth avoiding.  (Thread local access code is much worse...)

ghalliday commented on thorlcr/slave/thslavemain.cpp:
```diff
@@ -393,6 +393,7 @@ int main( int argc, const char *argv[]  )
 #else
         globals.setown(loadConfiguration(globals, argv, "thor", "THOR", nullptr, nullptr, nullptr, false));
 #endif
+        workunitGraphCacheEnabled = globals->getPropBool("expert/@workunitGraphCacheEnabled", workunitGraphCacheEnabled);
```
I think this is too early.  (See the comment in the code below....)
I would expect a similar call in the thor master. 

@jakesmith Does the slave load the graph from the workunit?  Does it reload it for each subgraph?

jakesmith replied:
Not per subgraph, but per Thor 'job' (i.e. each workflow graph)

jakesmith replied:
>I think this is too early. (See the comment in the code below....)

yes, would be good to move within RegisterSelf next to similar expert tests,
and so that's bare-metal compatible, should use getExpertOptBool, i.e.:

jakesmith replied:
>I would expect a similar call in the thor master.

in master, a getExpertOptBool call should go in CJobManager::execute

richardkchapman replied:
Fixed

jakesmith reviewed (commented):
@richardkchapman - please see additional comments re. Thor.

jakesmith reviewed (approved):
@richardkchapman - looks good.

ghalliday commented on common/workunit/workunit.cpp:
```diff
@@ -10411,8 +10411,11 @@ IPropertyTree * CLocalWUGraph::getXGMMLTreeRaw() const
     return p->getPropTree("xgmml");
 }
 
+bool workunitGraphCacheEnabled = false;
```
Should this default on in 8.4 and then off in 9.6 to help reduce the potential for unexpected regressions.  Then it could be tested on development system to verify the impact.

ghalliday reviewed (commented):
I think ok to merge.  remaining question is about the default.  I would opt for default the same in 9.4 and on in 9.6 via a separate PR. 

Please squash. Thanks.

richardkchapman commented:
I have changed the default to be true in this PR that targets 9.4, and squashed. Once merged and upmerged I can do a new PR to change the default in 9.6

ghalliday commented:
@mckellyln this option will be available in 9.4.x.  It would be worth testing it on the development systems once this build is deployed to see what effect it has on memory usage and performance.
